File: cpp-solutions/39.cpp
class Solution {
private:
    vector<vector<int>> res;
    vector<int> vec;
    int tar;
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        tar = target;
        helper(candidates, 0, 0);
        return res;
    }
    
    void helper(vector<int>& can, int s, int pos){
        for(int i=pos; i<can.size(); i++){
            int sum = s + can[i];
            if(sum == tar){
                vec.push_back(can[i]);
                res.push_back(vec);
                vec.pop_back();
                return;
            }else if(sum>tar){
                return;
            }else{
                vec.push_back(can[i]);
                helper(can, sum, i);
                vec.pop_back();
            }
        }
    }
};

File: golang/39/39.go
package main

import "sort"

var (
	res [][]int
	cur []int
)

func combinationSum(candidates []int, target int) [][]int {
	res = make([][]int, 0)
	cur = make([]int, 0)

	sort.Ints(candidates)
	backtrack(candidates, 0, 0, target)
	return res
}

func backtrack(nums []int, start, sum, target int) {
	if sum == target {
		t := make([]int, len(cur))
		copy(t, cur)
		res = append(res, t)
		return
	}

	if sum > target {
		return
	}

	for i := start; i < len(nums); i++ {
		cur = append(cur, nums[i])
		backtrack(nums, i, sum+nums[i], target)
		cur = cur[:len(cur)-1]
	}
}
